File: generated_code/python/solution_2347_2.py
from typing import Optional, Tuple

# Definition for a binary tree node if not already defined by the runner.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        self.matching_nodes_count = 0
        
        # Post-order traversal helper function.
        # Returns a tuple: (sum_of_subtree_values, number_of_nodes_in_subtree)
        def traverse(node: Optional[TreeNode]) -> Tuple[int, int]:
            if not node:
                return 0, 0
            
            # Recurse on left and right subtrees
            left_sum, left_count = traverse(node.left)
            right_sum, right_count = traverse(node.right)
            
            # Calculate current subtree metrics
            current_sum = left_sum + right_sum + node.val
            current_count = left_count + right_count + 1
            
            # Check if the node value equals the integer division average of the subtree
            if current_sum // current_count == node.val:
                self.matching_nodes_count += 1
                
            return current_sum, current_count
        
        traverse(root)
        return self.matching_nodes_count

File: generated_code/cpp/solution_2347_2.cpp
#include <algorithm>

// Definition for a binary tree node provided for compilability.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    // Helper structure to hold the accumulated sum and count of nodes in a subtree.
    struct SubtreeResult {
        int sum;
        int count;
    };

    int matchingNodesCount = 0;

    // Post-order traversal (Bottom-Up DP) to compute subtree sums and counts efficiently.
    // Time Complexity: O(N) where N is the number of nodes, as each node is visited once.
    // Space Complexity: O(H) where H is the height of the tree, due to the recursion stack.
    SubtreeResult calculateSubtree(TreeNode* node) {
        if (!node) {
            return {0, 0};
        }

        // Recursively find the sum and count of the left and right subtrees
        SubtreeResult leftResult = calculateSubtree(node->left);
        SubtreeResult rightResult = calculateSubtree(node->right);

        // Current subtree metrics include the current node's value
        int currentSum = leftResult.sum + rightResult.sum + node->val;
        int currentCount = leftResult.count + rightResult.count + 1;

        // Calculate average using integer division (automatically truncates towards zero/rounds down for non-negative values)
        if (currentSum / currentCount == node->val) {
            matchingNodesCount++;
        }

        return {currentSum, currentCount};
    }

public:
    int averageOfSubtree(TreeNode* root) {
        matchingNodesCount = 0;
        calculateSubtree(root);
        return matchingNodesCount;
    }
};